torch/csrc/jit/import.cpp

```diff
@@ -15,514 +15,308 @@ namespace torch { namespace jit {
-    , graph_reader(proto.graph) {}
-  Reader(pb_callback_t& cb)
-    : Reader() { initialize_callback(cb); }
-
-  virtual void decode(pb_istream_t *stream) override {
-    if (!pb_decode(stream, onnx_ModelProto_fields, &proto)) {
-      throw std::runtime_error("Decoding failed");
-    }
-
-    value.graph = std::move(graph_reader.value);
-  }
-
-  static Model_ read(pb_istream_t *stream) {
-    Reader<Model_> reader;
-    reader.decode(stream);
-    return reader.value;
-  }
-
-  onnx_ModelProto proto;
-  Reader<Graph_> graph_reader;
-  Model_ value;
-};
-
-
-// IR graph construction
-
-at::Tensor buildTensor(const Tensor_& tensor_) {
-
-  at::Tensor tensor;
-
-  switch(tensor_.data_type) {
-    case onnx_TensorProto_DataType_UINT8:
-      tensor = at::CPU(at::kByte).tensor();
-      break;
-    case onnx_TensorProto_DataType_INT8:
-      tensor = at::CPU(at::kChar).tensor();
-      break;
-    case onnx_TensorProto_DataType_INT16:
-      tensor = at::CPU(at::kShort).tensor();
-      break;
-    case onnx_TensorProto_DataType_INT32:
-      tensor = at::CPU(at::kInt).tensor();
-      break;
-    case onnx_TensorProto_DataType_INT64:
-      tensor = at::CPU(at::kLong).tensor();
-      break;
-    case onnx_TensorProto_DataType_FLOAT16:
-      tensor = at::CPU(at::kHalf).tensor();
-      break;
-    case onnx_TensorProto_DataType_FLOAT:
-      tensor = at::CPU(at::kFloat).tensor();
-      break;
-    case onnx_TensorProto_DataType_DOUBLE:
-      tensor = at::CPU(at::kDouble).tensor();
-      break;
+at::ScalarType JitDecoder::onnxTypeToATenType(onnx_torch::TensorProto_DataType onnx_type) {
+  switch(onnx_type) {
+    case onnx_torch::TensorProto_DataType_UINT8:
+      return at::kByte;
+    case onnx_torch::TensorProto_DataType_INT8:
+      return at::kChar;
+    case onnx_torch::TensorProto_DataType_INT16:
+      return at::kShort;
+    case onnx_torch::TensorProto_DataType_INT32:
+      return at::kInt;
+    case onnx_torch::TensorProto_DataType_INT64:
+      return at::kLong;
+    case onnx_torch::TensorProto_DataType_FLOAT16:
+      return at::kHalf;
+    case onnx_torch::TensorProto_DataType_FLOAT:
+      return at::kFloat;
+    case onnx_torch::TensorProto_DataType_DOUBLE:
+      return at::kDouble;
     default:
       throw std::runtime_error("Unsupported data type");
   }
-
-  tensor.resize_(tensor_.dims);
-
-  TORCH_ASSERT(tensor.storage()->size() * tensor.storage()->elementSize() == tensor_.raw_data.size());
-
-  std::memcpy(tensor.data_ptr(), tensor_.raw_data.data(), tensor_.raw_data.size());
-
-  return tensor;
 }
 
-Graph_ readSubgraph(const std::string& serialized_subgraph) {
-  pb_istream_t istream = pb_istream_from_buffer(reinterpret_cast<const pb_byte_t *>(serialized_subgraph.data()), serialized_subgraph.size());
-
-  return Reader<Graph_>::read(&istream);
+at::Tensor JitDecoder::buildTensor(const onnx_torch::TensorProto& tensor_proto) {
+  at::Tensor tensor = at::CPU(onnxTypeToATenType(tensor_proto.data_type())).tensor();
```

Comment:
btw, should we store also tensor's device? or things magically turn to cpu after serialization?

Reply:
I was thinking that everything should turn into a CPU tensor, because loading the module will mmap the storage from the file directly, but maybe it is a better idea to keep all the information and then only use mmap for the CPU tensors.